### PR TITLE
Convert RefreshSessionIfNeeded into RefreshSession

### PR DESCRIPTION
**Pull request by NickMeves**
https://github.com/oauth2-proxy/oauth2-proxy/issues/856

As touched on in the Issue & in this CSRF PR: https://github.com/oauth2-proxy/oauth2-proxy/pull/967 - We now validate after any redeem or refresh attempt. The rationale being some providers might want to validate the response from those endpoints (e.g. nonce claim for security).

Plus, now that refreshSession can happen early, if it fails - we don't fail and delete the session, we do a validation check to see if the existing session was still valid.

The old validate logic where we only called it if `RefreshSessionIfNeeded` (making it only run it refresh timer was shorter than provider token TTLs) doesn't make sense anymore now that we force refreshes early.

## How Has This Been Tested?

Unit Tests

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: pkg/middleware/stored_session.go [hunk 1]
```diff
@@ -108,49 +108,47 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
...
+		return fmt.Errorf("error refreshing tokens: %v", err)
 	}
 
 	if !refreshed {
-		return false, nil
+		return nil
 	}
 
+	// If we refreshed, update the `CreatedAt` time to reset the refresh timer
+	session.CreatedAtNow()
```
Comment (NickMeves) [inline pkg/middleware/stored_session.go:143]: I wanted to make this explicit in our framework rather than relying on provider implementation since we've had provider PRs forget to reset this on refresh before.

File: providers/azure.go [hunk 2]
```diff
@@ -146,40 +146,41 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (*
...
 		return false, fmt.Errorf("unable to redeem refresh token: %v", err)
 	}
 
-	logger.Printf("refreshed id token %s (expired on %s)\n", s, origExpiration)
 	return true, nil
 }
 
-func (p *AzureProvider) redeemRefreshToken(ctx context.Context, s *sessions.SessionState) (err error) {
+func (p *AzureProvider) redeemRefreshToken(ctx context.Context, s *sessions.SessionState) error {
+	clientSecret, err := p.GetClientSecret()
```
Comment (NickMeves) [inline providers/azure.go:176]: Unrelated, but I noticed Azure was missing this support so I added it while cleaning things up.

File: providers/provider_default.go [hunk 3]
```diff
@@ -128,8 +131,14 @@ func (p *ProviderData) ValidateSession(ctx context.Context, s *sessions.SessionS
...
-func (p *ProviderData) RefreshSessionIfNeeded(_ context.Context, _ *sessions.SessionState) (bool, error) {
-	return false, nil
+func (p *ProviderData) RefreshSession(_ context.Context, s *sessions.SessionState) (bool, error) {
+	if s == nil {
+		return false, nil
+	}
+
+	// Pretend `RefreshSession` occurred so `ValidateSession` isn't called
+	// on every request after any potential set refresh period elapses.
+	return true, nil
```
Comment (NickMeves) [inline providers/provider_default.go:141]: This is for our few providers that implemented a `ValidateSession` but not `RefreshSession` -- I think in the current implementation if any users set a `--cookie-refresh`, once that time had elapsed, every subsequent request would hit the validate endpoint because they could never successfully refresh and reset the `CreatedAt` time.

File: pkg/apis/sessions/session_state.go [hunk 4]
```diff
@@ -30,9 +33,30 @@ type SessionState struct {
 	PreferredUsername string   `msgpack:"pu,omitempty"`
 }
 
+// CreatedAtNow sets a SessionState's CreatedAt to now
```
Comment (NickMeves) [inline pkg/apis/sessions/session_state.go:36]: Open to suggestions on this.

This was all over the place in the providers with embedded direct `time.Now()` calls making session timing logic within them impossible to mock effectively.

I want to try to take more control of setting the session pieces here in a uniform way instead of stuff just dumping fields into the public struct. Thoughts?

File: pkg/apis/sessions/session_state.go [hunk 5]
```diff
@@ -30,9 +33,30 @@ type SessionState struct {
...
+}
+
+// SetExpiresOn sets an expiration
+func (s *SessionState) SetExpiresOn(exp time.Time) {
+	s.ExpiresOn = &exp
+}
+
+// ExpiresIn sets an expiration a certain duration from CreatedAt.
+// CreatedAt will be set to time.Now if it is unset.
+func (s *SessionState) ExpiresIn(d time.Duration) {
```
Comment (JoelSpeed) [inline pkg/apis/sessions/session_state.go:49]: The naming of this implied to me it would be expiring the duration after now, not duration after create, why do we want this to be expire after create rather than expire after now?

File: pkg/middleware/stored_session.go [hunk 6]
```diff
@@ -108,49 +108,47 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
...
+		logger.Errorf("Unable to refresh session: %v", err)
 	}
 
-	if !refreshed {
-		// Session wasn't refreshed, so make sure it's still valid
-		return s.validateSession(req.Context(), session)
-	}
-	return nil
+	// Validate all sessions after any Redeem/Refresh operation (fail or success)
+	return s.validateSession(req.Context(), session)
```
Comment (JoelSpeed) [inline pkg/middleware/stored_session.go:127]: Session validation in some instances calls the API/token endpoint, this would cause an increase in the number of times we are hitting the API, anything we can do about this?

File: pkg/middleware/stored_session.go (hunk 1 above)
Comment (JoelSpeed) [reply pkg/middleware/stored_session.go:143]: Do we need to also reset it if the refresh logic isn't enabled/implemented? So we just update the cookie if it's still valid?

File: providers/provider_default.go (hunk 3 above)
Comment (JoelSpeed) [reply providers/provider_default.go:141]: Ok I think this answers my previous question about what happens for those that don't implement the refresh. I wonder if instead we should return an explicity `ErrNotImplemented` and handle that? It would be more obvious to future contributors IMO

File: pkg/apis/sessions/session_state.go (hunk 5 above)
Comment (NickMeves) [reply pkg/apis/sessions/session_state.go:49]: I was torn on this implementation -- in practicality, this is consumed by OAuth redeem responses that have the `expires_in` field (analogous to the other version that had explicit `ExpiresOn`).

They immediately create a fresh session, then set the expires to the that expires in duration. I guess the time drift between these multiple calls is practically negligible, I just wanted to perfectly align the TTL difference between createdAt/expiresOn to equal the `expires_in`.

File: pkg/apis/sessions/session_state.go (hunk 4 above)
Comment (NickMeves) [reply pkg/apis/sessions/session_state.go:36]: So sessions is where I wanted to talk about our time mocking posture more in general. These are consumed throughout a bunch of our other packages. For them to more accurately test session expiration scenarios, we likely need an entry point to fast forward session time (without access to the underlying session because it might be internal to the other package we might be consuming - e.g. session storage, cookies).

Comment (NickMeves) [reply pkg/apis/sessions/session_state.go:36]: Here's the brain dump I put into Slack a bit ago:

I want to talk more about redesigning how we use time.Now in the project and how we expose it in a convenient way that allows for good tests (we’re probably lacking on border expiration unit test cases because we can’t mock time easily in some areas).

Options: now time.Time method parameters (likely adds code climate violations), nowFunc func() time.Time fields on structs (a little bloat for widely used objects, only available within package tests due to private), var nowFunc = time.Time private package variables (again only available within package tests), or var NowFunc = time.Time public package variables (a little sloppy exposing this, but gives the best coverage for testing — particularly integration tests we are missing from the project where we want to fully mock the flow of time).

Want me to throw together an RFC issue so we can discuss this more publicly? After integrating mockoidc with jwt-go I am starting to lean heavily toward the last one — because they exposed time.Now publically to override, I could completely mock time and test border cases for JWT expiration & refresh scenarios in end to end integration tests.

File: pkg/middleware/stored_session.go (hunk 6 above)
Comment (NickMeves) [reply pkg/middleware/stored_session.go:127]: Yeah, that was my concern about this too 🤔 

Not sure on that one, some providers send a dummy request to an endpoint to verify the access token (which probably isn't needed if a refresh just granted the new token -- but maybe doesn't hurt to make sure it work?).

The other case we will have more of in OIDC providers is verifying security-based IDToken claims - so their implementation won't have the API request.

In practice at the moment, the only provider's that implemented an API reqeust Validation didn't implement a RefreshSession function, so nothing currently has a double hit 🤷‍♂️

File: pkg/apis/sessions/session_state.go (hunk 4 above)
Comment (NickMeves) [reply pkg/apis/sessions/session_state.go:36]: > My thinking here is that you ensure that the NowFunc only affects the current session you're manipulating in the tests, other tests could be run in parallel etc without worry

Ah parallel tests... didn't even think about that. Yes, that makes package level time functions a much less desirable way to manage this 🤔

**Comment by Bibob7** [comment]
I would offer my help here, because this is still relevant for me and prerequirement to use distributed locks.

**Comment by NickMeves** [comment]
@JoelSpeed I'll want to give this another review myself since its been 3 months since I originally started this -- but on an initial rebase & finishing touches, this seems good for a review.

File: pkg/apis/sessions/session_state.go [hunk 7]
```diff
@@ -235,35 +251,3 @@ func lz4Decompress(compressed []byte) ([]byte, error) {
...
-//
-// Non-empty check is needed due to ensure the non-authenticated AES-CFB
-// decryption doesn't result in garbage data that collides with a valid
-// MessagePack header bytes (which MessagePack will unmarshal to an empty
-// default SessionState). <1% chance, but observed with random test data.
-//
-// UTF-8 check ensures the strings are valid and not raw bytes overloaded
-// into Latin-1 encoding. The occurs when legacy unencrypted fields are
-// decrypted with AES-CFB which results in random bytes.
-func (s *SessionState) validate() error {
```
Comment (NickMeves) [inline pkg/apis/sessions/session_state.go:249]: Now that Locks & Clocks are in the Session for internal usage and not all fields are serialized, the `reflect.DeepEqual` might have issues moving forward (in earlier iterations, it did not like the `Clock` mutex).

This really isn't needed anymore and is an unneeded computation in the session deserialization that happens every request. I added an important note on the migration considerations for the slim user base this might affect.

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,9 +4,14 @@
 
 ## Important Notes
 
+- [#1086](https://github.com/oauth2-proxy/oauth2-proxy/pull/1086) The extra validation to protect invalid session deserialization
```
Comment (NickMeves) [inline CHANGELOG.md:7]: Rationale:

SHA256 was added in v6.0.0
LZ4 + MessagePack was added in v6.1.0

Underlying encryption is still AES-CFB which is unauthenticated where the SHA is our MAC. A session created in v6.0.0 could be SHA256 MAC, then CFB decrypt, and old style JSON would possible pass LZ4 decompression & messagepack might see leading bytes and unmarshal an empty `Session` instead of erroring.

File: pkg/clock/clock.go [hunk 9]
```diff
@@ -63,13 +63,10 @@ func Reset() *clockapi.Mock {
 // package.
 type Clock struct {
 	mock *clockapi.Mock
-	sync.Mutex
```
Comment (NickMeves) [inline pkg/clock/clock.go:66]: The underlying mutex was causing issues with the linter when any `Session` objects were copied by value. This pattern was used heavily in tests, so it was a real hassle.

Since this is the mutex on the local Clock instance (not global) and didn't even apply to time operations, I'm not sure how exactly test code could footgun itself by running parallel Set/Add/Reset -- I don't see a pattern where on a localized Clock, it would somehow be parallized but test management code would be fast-forwarding a single local in parallel? Seems like an anti-pattern if it were tried 🤷‍♂️ 

And the underlying `clockapi.Mock` manages itself with locks throughout.

File: pkg/middleware/stored_session.go (hunk 6 above)
Comment (NickMeves) [reply pkg/middleware/stored_session.go:127]: This is still an issue.

Is it worth checking on Google/Azure to see if `ValidateSession` can be refactored to check the IDToken validity like OIDC instead of a Profile URL network request?

Or is 2 IdP round trips instead of 1 on refreshes not too big of a deal?

CC: @weinong @codablock

File: pkg/apis/sessions/session_state.go (hunk 4 above)
Comment (NickMeves) [reply pkg/apis/sessions/session_state.go:66]: Handled in Clock package now, resolving.

File: providers/providers.go [hunk 10]
```diff
@@ -9,14 +9,14 @@ import (
 // Provider represents an upstream identity provider implementation
 type Provider interface {
 	Data() *ProviderData
+	GetLoginURL(redirectURI, finalRedirect string, nonce string) string
```
Comment (NickMeves) [inline providers/providers.go:12]: Quality of life -- ordered the interface methods in their OAuth flow order.

**Comment by NickMeves** [comment]
> I would offer my help here, because this is still relevant for me and prerequirement to use distributed locks.

@Bibob7 Do you mind testing this interactively in your environment to see if it starts to satisfy your requirements for finishing the refresh lock work?

File: pkg/apis/sessions/session_state.go (hunk 5 above)
Comment (JoelSpeed) [reply pkg/apis/sessions/session_state.go:79]: Would `ExpiresAfter` make it any more obvious? :thinking: 

I get what you're trying to do here and don't have an issue with the logic, just wish we could come up with a more obvious name

File: pkg/middleware/stored_session.go (hunk 6 above)
Comment (JoelSpeed) [reply pkg/middleware/stored_session.go:127]: I'm not sure what to do on this one, if it is just the two providers, perhaps we can look into deduplicating it, but  you're right, 2 calls may not be the end of the world

File: providers/provider_default.go (hunk 3 above)
Comment (JoelSpeed) [reply providers/provider_default.go:145]: I'm confused about why we aren't using a special error here? Why will all providers call this? Surely the error will be passed up to the session middleware which can then handle it?

Comment (NickMeves) [reply providers/provider_default.go:145]: I'll take another peek and see if we can implement with the error, I gave it a whirl and it somehow felt tedious, but I can't remember why 🤔 

The other spots of the interface where we have `ErrNotImplemented` are truly optional codepaths (e.g. `CreateSessionFromToken`) where if someone tried to send a bearer token to a provider that didn't support it, it would error out.

Vs `ValidateSession` which is in the login (and soon refresh) flow for all providers.

File: pkg/middleware/stored_session.go [hunk 11]
```diff
@@ -108,49 +109,57 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
...
+// Success or fail, we will then validate the session.
 func (s *storedSessionLoader) refreshSessionIfNeeded(rw http.ResponseWriter, req *http.Request, session *sessionsapi.SessionState) error {
 	if s.refreshPeriod <= time.Duration(0) || session.Age() < s.refreshPeriod {
 		// Refresh is disabled or the session is not old enough, do nothing
 		return nil
 	}
 
-	logger.Printf("Refreshing %s old session cookie for %s (refresh after %s)", session.Age(), session, s.refreshPeriod)
-	refreshed, err := s.refreshSessionWithProvider(rw, req, session)
+	logger.Printf("Refreshing session - User: %s; SessionAge: %s", session.User, session.Age())
```
Comment (NickMeves) [inline pkg/middleware/stored_session.go:119]: Scope Creep: Removing the Email from the logs for our GDPR constrained users.

File: pkg/apis/sessions/session_state.go (hunk 5 above)
Comment (NickMeves) [reply pkg/apis/sessions/session_state.go:79]: I was thinking making it exactly `ExpiresIn` signifies it should be used with any OAuth providers that use the `expires_in` field in their JSON response.  Where the other `SetExpiresOn` (naming conflict with `ExpiresOn` field) should be used with any OAuth providers that opt for `expires_on` in their response schema.

1:1 alignment with the OAuth schema was my goal for clarity when people implement (I agree, the name doesn't fully reflect what the logic does 😅 )

File: pkg/middleware/stored_session.go (hunk 6 above)
Comment (NickMeves) [reply pkg/middleware/stored_session.go:128]: Ideally, we get Azure & Google as OIDC based providers long term. Then we validate with an IDToken signature verification instead of a profileURL API call.

File: providers/provider_default.go (hunk 3 above)
Comment (NickMeves) [reply providers/provider_default.go]: Latest commit has the shift 👍

File: pkg/middleware/stored_session.go [hunk 12]
```diff
@@ -108,49 +109,57 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
...
 	if !refreshed {
-		return false, nil
+		if errors.Is(err, providers.ErrNotImplemented) {
+			// HACK:
+			// Providers that don't implement `RefreshSession` use the default
+			// implementation which returns `ErrNotImplemented`.
+			// Update `session.CreatedAt` to reset the refresh timer so that
+			// `ValidateSession` isn't triggered every subsequent request and is
+			// only called once during this request.
+			session.CreatedAtNow()
```
Comment (JoelSpeed) [inline pkg/middleware/stored_session.go:147]: Since we modified the session, do we not now need to save the session? Or do we actually just internally update the `createdAt`? Does this prevent that dreaded refreshing session log message that happens in this case?

**Review by JoelSpeed — COMMENTED** [review]
I think I'm good with this now apart from one minor bit about the sessions that don't implement refresh. It was my understanding that we would want to update the created at for real but we don't seem to be saving the session?

File: pkg/middleware/stored_session.go (hunk 6 above)
Comment (JoelSpeed) [reply pkg/middleware/stored_session.go:128]: What do you want to do then with regards to this PR for now?

File: pkg/middleware/stored_session.go (hunk 12 above)
Comment (NickMeves) [reply pkg/middleware/stored_session.go:147]: Oh, you are totally right! I lost the saving aspect trying to flip to `ErrNotImplemented`, let me fix that up.

Comment (NickMeves) [reply pkg/middleware/stored_session.go]: Latest commit has the fix. Added some test cases too since the old cases I had only followed the hack where I pretended the default case was implemented and returned true.

File: pkg/middleware/stored_session.go [hunk 13]
```diff
@@ -108,49 +109,56 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
...
+	// that `ValidateSession` isn't triggered every subsequent request and is
+	// only called once during this request.
+	if !errors.Is(err, providers.ErrNotImplemented) {
+		if err != nil {
+			return fmt.Errorf("error refreshing tokens: %v", err)
+		}
+		// Session not refreshed, nothing to persist.
+		if !refreshed {
+			return nil
+		}
```
Comment (JoelSpeed) [inline pkg/middleware/stored_session.go:148]: For some reason I'm really struggling to follow how and why this is correct :thinking: 

I'm wondering if there is a different hacky way to make this easier to understand

Is this the same as 
```
if err !=nil && !errors.Is(err, providers.ErrNotImplemented) {
   return err
}

if errors.Is(err, providers.ErrNotImplemented) {
    // Insert comment about this being a hack here
    refreshed = true
}

if !refreshed {
    return nil
}
```

**Review by JoelSpeed — COMMENTED** [review]
One little nit about me finding this hard to read, 5/10 on this, lmk what you think

File: pkg/middleware/stored_session.go (hunk 13 above)
Comment (NickMeves) [reply pkg/middleware/stored_session.go:148]: Yeah - this looks like it unnests it well to make it clearer. (I originally had something similar, but consolidated the `ErrNotImplemented` check since it was happening multiple times BUT at the expense of more nesting).

This is getting back to the original design though -- just having the default implementation return `true` for `refreshed` even though it didn't do anything.

Want to leave that as `false` in the implementation but override here at the source of the need for the hack with the explanation?

Comment (JoelSpeed) [reply pkg/middleware/stored_session.go]: > Want to leave that as false in the implementation but override here at the source of the need for the hack with the explanation?

I think this makes the most sense, it's up to the caller to determine how they handle a not implemented in most cases I see. So i think it makes sense to have the hack description right in the logic here, rather than in the provider implementation where we might not see/understand it as we don't have the context

**Review by JoelSpeed — APPROVED** [review]
If you want to merge this, I think it's good to go now. Quite a few commits, did you want to squash down at all?
